README.md
[![Build Status](https://travis-ci.org/kazeburo/WWW-Form-UrlEncoded.png?branch=master)](https://travis-ci.org/kazeburo/WWW-Form-UrlEncoded)
# NAME

WWW::Form::UrlEncoded - parser and builder for application/x-www-form-urlencoded

# SYNOPSIS

    use WWW::Form::UrlEncoded qw/parse_urlencoded build_urlencoded/;
    

    my $query_string = "foo=bar&baz=param";
    my @params = parse_urlencoded($query_string);
    # ('foo','bar','baz','param')
    

    my $query_string = build_urlencoded('foo','bar','baz','param');
    # "foo=bar&baz=param";

# DESCRIPTION

WWW::Form::UrlEncoded provides application/x-www-form-urlencoded parser and builder.
This module aims to have compatibility with other CPAN modules like 
HTTP::Body's urlencoded parser.

This module try to use [WWW::Form::UrlEncoded::XS](http://search.cpan.org/perldoc?WWW::Form::UrlEncoded::XS) by default and fail to it, 
use WWW::Form::UrlEncoded::PP instead

## Parser rules

WWW::Form::UrlEncoded parsed string in this rule.

1. Split application/x-www-form-urlencoded payload by `&` (U+0026) or `;` (U+003B)
2. Ready empty array to store `name` and `value`
3. For each divided string, apply next steps.
    1. If first character of string is `' '` (U+0020 SPACE), remove it.
    2. If string has `=`, let __name__ be substring from start to first `=`, but excluding first `=`, and remains to be __value__. If there is no strings after first `=`, __value__ to be empty string `""`. If first `=` is first character of the string, let __key__ be empty string `""`. If string does not have any `=`, all of the string to be __key__ and __value__ to be empty string `""`.
    3. replace all `+` (U+002B) with `' '` (U+0020 SPACE).
    4. unescape __name__ and __value__. push them to the array.
4. return the array.

## Test data

    'a=b&c=d'     => ["a","b","c","d"]
    'a=b;c=d'     => ["a","b","c","d"]
    'a=1&b=2;c=3' => ["a","1","b","2","c","3"]
    'a==b&c==d'   => ["a","=b","c","=d"]
    'a=b& c=d'    => ["a","b","c","d"]
    'a=b; c=d'    => ["a","b","c","d"]
    'a=b; c =d'   => ["a","b","c ","d"]
    'a=b;c= d '   => ["a","b","c"," d "]
    'a=b&+c=d'    => ["a","b"," c","d"]
    'a=b&+c+=d'   => ["a","b"," c ","d"]
    'a=b&c=+d+'   => ["a","b","c"," d "]
    'a=b&%20c=d'  => ["a","b"," c","d"]
    'a=b&%20c%20=d' => ["a","b"," c ","d"]
    'a=b&c=%20d%20' => ["a","b","c"," d "]
    'a&c=d'       => ["a","","c","d"]
    'a=b&=d'      => ["a","b","","d"]
    'a=b&='       => ["a","b","",""]
    '&'           => ["","","",""]
    '='           => ["",""]
    ''            => []

# FUNCTION

- @param = parse\_urlencoded($str:String)

    parse `$str` and return Array that contains key-value pairs.

- $string = build\_urlencoded(@param)
- $string = build\_urlencoded(@param, $delim)
- $string = build\_urlencoded(\\@param)
- $string = build\_urlencoded(\\@param, $delim)
- $string = build\_urlencoded(\\%param)
- $string = build\_urlencoded(\\%param, $delim)

    build urlencoded string from __param__. build\_urlencoded accepts arrayref and hashref values.

        build_urlencoded( foo => 1, foo => 2);
        build_urlencoded( foo => [1,2] );
        build_urlencoded( [ foo => 1, foo => 2 ] );
        build_urlencoded( [foo => [1,2]] );
        build_urlencoded( {foo => [1,2]} );

    If `$delim` parameter is passed, this function use it instead of using `&`.

- $string = build\_urlencoded\_utf8(...)

    This function is almost same as `build_urlencoded`. build\_urlencoded\_utf8 call `utf8::encode` for all parameters.

# ENVIRONMENT VALUE

- WWW\_FORM\_URLENCODED\_PP

    If true, WWW::Form::UrlEncoded force to load WWW::Form::UrlEncoded::PP.

# SEE ALSO

CPAN already has some application/x-www-form-urlencoded parser modules like these.

- [URL::Encode](http://search.cpan.org/perldoc?URL::Encode)
- [URL::Encode::XS](http://search.cpan.org/perldoc?URL::Encode::XS)
- [Text::QueryString](http://search.cpan.org/perldoc?Text::QueryString)

They does not fully compatible with WWW::Form::UrlEncoded. Handling of empty key-value
and supporting separator characters are different.

# LICENSE

Copyright (C) Masahiro Nagano.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

Masahiro Nagano <kazeburo@gmail.com>


Changes
Revision history for Perl extension WWW-Form-UrlEncoded

{{$NEXT}}

0.23 2015-11-19T00:33:17Z

   - bugfix. parse(undef) return arrayref

0.22 2015-06-12T05:37:24Z

   - \C is deprecated at perl-5.22 (Thank you eserte)

0.21 2015-04-28T17:04:09Z

   - add some tests related https://github.com/kazeburo/WWW-Form-UrlEncoded-XS/issues/1

0.19 2014-04-30T03:26:43Z

   - bugfix: PP::build_urlencoded should not do utf8::upgrade.

0.17 2014-02-24T13:50:55Z

    - fix EXPORT bug. _arrayref did not export

0.16 2014-02-24T08:14:53Z

    - add parse_urlencoded_arrayref

0.14 2014-02-21T04:21:03Z

    - add build_urlencoded_utf8

0.13 2014-02-18T07:59:59Z

    - supports a delimiter parameter

0.12 2014-02-18T03:38:24Z

    - install .keep file under $arch/WWW/Form/UrlEncoded/XS to keep compatiblity before 0.10
    - build_urlencoded supports arrayref and hashref args

0.10 2014-02-17T06:10:58Z

    - [IMPORTANT] split WWW::Form::UrlEncoded into another dist.
                  instal with `cpanm WWW::Form::UrlEncoded` 

0.04 2014-02-17T03:30:50Z

    - add build_urlencoded

0.03 2014-02-10T16:10:56Z

    - fix bug with thread enabled perl

0.02 2014-02-10T02:25:59Z

    - mv ppport.h to src

0.01 2014-02-07T05:44:01Z

    - original version



Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;
use CPAN::Meta;
use CPAN::Meta::Prereqs;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'WWW-Form-UrlEncoded',
    module_name     => 'WWW::Form::UrlEncoded',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],
    c_source     => [qw()],
    PL_files => {},

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,

    
);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

my $mbmeta = CPAN::Meta->load_file('MYMETA.json');
my $meta = CPAN::Meta->load_file('META.json');
my $prereqs_hash = CPAN::Meta::Prereqs->new(
    $meta->prereqs
)->with_merged_prereqs(
    CPAN::Meta::Prereqs->new($mbmeta->prereqs)
)->as_string_hash;
my $mymeta = CPAN::Meta->new(
    {
        %{$meta->as_struct},
        prereqs => $prereqs_hash
    }
);
print "Merging cpanfile prereqs to MYMETA.yml\n";
$mymeta->save('MYMETA.yml', { version => 1.4 });
print "Merging cpanfile prereqs to MYMETA.json\n";
$mymeta->save('MYMETA.json', { version => 2 });


eg/bench.pl
#!/usr/bin/perl

use strict;
use warnings;
use Benchmark qw/:all/;
use WWW::Form::UrlEncoded;
use WWW::Form::UrlEncoded::PP;
use URL::Encode::XS;
use Text::QueryString;


my @query_string = (
    "foo=bar",
    "foo=bar&bar=1",
    "foo=bar;bar=1",
    "foo=bar&foo=baz",
    "foo=bar&foo=baz&bar=baz",
    "foo_only",
    "foo&bar=baz",
    '%E6%97%A5%E6%9C%AC%E8%AA%9E=%E3%81%AB%E3%81%BB%E3%82%93%E3%81%94&%E3%81%BB%E3%81%92%E3%81%BB%E3%81%92=%E3%81%B5%E3%81%8C%E3%81%B5%E3%81%8C',
);

my $xs = Text::QueryString->new;

cmpthese(timethese(-1, {
    text_qs => sub {
        foreach my $qs (@query_string) {
            my @q = $xs->parse($qs);
        }
    },
    wwwform_xs => sub {
        foreach my $qs (@query_string) {
            my @q = WWW::Form::UrlEncoded::XS::parse_urlencoded($qs);
        }
    },
    wwwform_pp => sub {
        foreach my $qs (@query_string) {
            my @q = WWW::Form::UrlEncoded::PP::parse_urlencoded($qs);
        }
    },
    urlencode_xs => sub {
        foreach my $qs (@query_string) {
            my @q = @{URL::Encode::XS::url_params_flat($qs)};
        }
    },
}));


__END__
Benchmark: running text_qs, urlencode_xs, wwwform_pp, wwwform_xs for at least 1 CPU seconds...
   text_qs:  1 wallclock secs ( 1.08 usr +  0.00 sys =  1.08 CPU) @ 53096.30/s (n=57344)
urlencode_xs:  1 wallclock secs ( 1.08 usr +  0.00 sys =  1.08 CPU) @ 66369.44/s (n=71679)
wwwform_pp:  1 wallclock secs ( 1.06 usr +  0.01 sys =  1.07 CPU) @ 11821.50/s (n=12649)
wwwform_xs:  1 wallclock secs ( 1.05 usr +  0.00 sys =  1.05 CPU) @ 91021.90/s (n=95573)
                Rate   wwwform_pp      text_qs urlencode_xs   wwwform_xs
wwwform_pp   11821/s           --         -78%         -82%         -87%
text_qs      53096/s         349%           --         -20%         -42%
urlencode_xs 66369/s         461%          25%           --         -27%
wwwform_xs   91022/s         670%          71%          37%           --




eg/bench_build.pl
#!/usr/bin/perl

use strict;
use warnings;
use Benchmark qw/:all/;
use URI;
use URI::Escape;
use URL::Encode::XS qw/url_encode/;
use WWW::Form::UrlEncoded::XS qw/build_urlencoded/;

my $base = q!http://api.example.com!;
my $path = q!/path/to/endpoint!;
my %param = (
    token => 'b a r',
    message => 'foo bar baz hoge hoge hoge hoge hogehoge',
    hoge => '日本語ですよー',
    arrayref => [qw/1 2 5/]
);

my @qs = (
    s_id => 1,
    type => 'foo',
    %param
);
warn build_urlencoded(@qs);

cmpthese(timethese(-2, {
    'use_uri' => sub {
        my $uri = URI->new($base . $path);
        $uri->query_form(
            s_id => 1,
            type => 'foo',
            %param
        );
        $uri->as_string;
    },
    'concat_uri' => sub {
        my @qs = (
            s_id => 1,
            type => 'foo',
            %param
        );
        my $uri = $base . $path . '?';
        while ( @qs ) {
            my $k = shift @qs;
            my $v = shift @qs;
            if ( ref $v && ref $v eq 'ARRAY') {
                $uri .= uri_escape($k) . '='. uri_escape($_) . '&' for @$v;
            }
            else {
                $uri .= uri_escape($k) . '='. uri_escape($v) . '&'
            }
        }
        substr($uri,-1,1,"");
        $uri;

    },
    'concat_url_encode_xs' => sub {
        my @qs = (
            s_id => 1,
            type => 'foo',
            %param
        );
        my $uri = $base . $path . '?';
        while ( @qs ) {
            my $k = shift @qs;
            my $v = shift @qs;
            if ( ref $v && ref $v eq 'ARRAY') {
                $uri .= url_encode($k) . '='. url_encode($_) . '&' for @$v;
            }
            else {
                $uri .= url_encode($k) . '='. url_encode($v) . '&'
            }
        }
        substr($uri,-1,1,"");
        $uri;
    },
    'build_urlencoded_xs' => sub {
        my $uri = $base . $path . '?' . build_urlencoded(
            s_id => 1,
            type => 'foo',
            %param
        );
        $uri;
    },
}));

__END__
Benchmark: running build_urlencoded_xs, concat_uri, concat_url_encode_xs, use_uri for at least 2 CPU seconds...
build_urlencoded_xs:  2 wallclock secs ( 2.08 usr +  0.00 sys =  2.08 CPU) @ 482635.58/s (n=1003882)
concat_uri:  3 wallclock secs ( 2.21 usr +  0.00 sys =  2.21 CPU) @ 19460.18/s (n=43007)
concat_url_encode_xs:  3 wallclock secs ( 2.12 usr +  0.00 sys =  2.12 CPU) @ 119165.09/s (n=252630)
   use_uri:  2 wallclock secs ( 2.17 usr +  0.00 sys =  2.17 CPU) @ 12387.10/s (n=26880)
                         Rate use_uri concat_uri concat_url_encode_xs build_urlencoded_xs
use_uri               12387/s      --       -36%                 -90%                -97%
concat_uri            19460/s     57%         --                 -84%                -96%
concat_url_encode_xs 119165/s    862%       512%                   --                -75%
build_urlencoded_xs  482636/s   3796%      2380%                 305%                  --


lib/WWW/Form/UrlEncoded/PP.pm
package WWW::Form::UrlEncoded::PP;

use strict;
use warnings;
use base qw/Exporter/;

our @EXPORT_OK = qw/parse_urlencoded parse_urlencoded_arrayref build_urlencoded build_urlencoded_utf8/;

our $DECODE = qr/%([0-9a-fA-F]{2})/;
our %DecodeMap;
our %EncodeMap;
for my $num ( 0 .. 255 ) {
    my $h = sprintf "%02X", $num;
    my $chr = chr $num;
    $DecodeMap{ lc $h } = $chr; #%aa
    $DecodeMap{ uc $h } = $chr; #%AA
    $DecodeMap{ ucfirst lc $h } = $chr; #%Aa
    $DecodeMap{ lcfirst uc $h } = $chr; #%aA
    $EncodeMap{$chr} = '%'. uc $h;
}
$EncodeMap{" "} = '+';

sub parse_urlencoded {
    return [] unless defined $_[0];
    my @params;
    for my $pair ( split( /[&;] ?/, $_[0], -1 ) ) {
        $pair =~ y/\+/\x20/;
        my ($key, $val) = split /=/, $pair, 2;
        for ($key, $val) {
            if ( ! defined $_ ) { 
                push @params, '';
                next;
            }
            s/$DECODE/$DecodeMap{$1}/gs;
            push @params, $_;
        }
    }

    return @params;
}

sub parse_urlencoded_arrayref {
    [parse_urlencoded(@_)];
}

our $NEED_UPGRADE = 1;
sub build_urlencoded {
    return "" unless @_;
    my $uri = '';
    my $delim = '&';
    if ( ref $_[0] && ref $_[0] eq 'ARRAY') {
        my @args = @{$_[0]};
        $delim = $_[1] if defined $_[1];
        utf8::encode($delim) if $NEED_UPGRADE;
        while ( @args ) {
            my $k = shift @args;
            my $v = shift @args;
            if ( ref $v && ref $v eq 'ARRAY') {
                $uri .= url_encode($k) . '='. url_encode($_) . $delim for @$v;
            }
            else {
                $uri .= url_encode($k) . '='. url_encode($v) . $delim
            }
        }
    }
    elsif ( ref $_[0] && ref $_[0] eq 'HASH') {
        $delim = $_[1] if defined $_[1];
        utf8::encode($delim) if $NEED_UPGRADE;
        while ( my ($k,$v) = each %{$_[0]} ) {
            if ( ref $v && ref $v eq 'ARRAY') {
                $uri .= url_encode($k) . '='. url_encode($_) . $delim for @$v;
            }
            else {
                $uri .= url_encode($k) . '='. url_encode($v) . $delim
            }
        }
    }
    else {
        if ( @_ > 2 && @_ % 2 ) {
            $delim = pop @_;
            utf8::encode($delim) if $NEED_UPGRADE;
        }
        while ( @_ ) {
            my $k = shift @_;
            my $v = shift @_;
            if ( ref $v && ref $v eq 'ARRAY') {
                $uri .= url_encode($k) . '='. url_encode($_) . $delim for @$v;
            }
            else {
                $uri .= url_encode($k) . '='. url_encode($v) . $delim
            }
        }
    }
    substr($uri,-1*length($delim),length($delim),"");
    $uri;
}

sub build_urlencoded_utf8 {
    local $NEED_UPGRADE = 1;
    my $uri = build_urlencoded(@_);
    $uri;
}

sub url_encode {
    return '' unless defined $_[0];
    my $t = shift;
    utf8::encode($t) if $NEED_UPGRADE;
    $t =~ s!([^A-Za-z0-9\-\._~])!
        join '',@EncodeMap{exists $EncodeMap{$1} ? ($1) : ($1 =~ /(\C)/gs)}
    !gsxe;
    return $t;
}

1;

__END__

=encoding utf-8

=head1 NAME

WWW::Form::UrlEncoded::PP - pure-perl parser and builder for application/x-www-form-urlencoded

=head1 SYNOPSIS

    use WWW::Form::UrlEncoded::PP qw/parse_urlencoded build_urlencoded/;
    
    my $query_string = "foo=bar&baz=param";
    my @params = parse_urlencoded($query_string);
    # ('foo','bar','baz','param')
    
    my $query_string = build_urlencoded('foo','bar','baz','param');
    # "foo=bar&baz=param";

=head1 DESCRIPTION

WWW::Form::UrlEncoded::PP provides pure-perl application/x-www-form-urlencoded parser and builder.
see L<WWW::Form::UrlEncoded>'s document.

=head1 LICENSE

Copyright (C) Masahiro Nagano.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Masahiro Nagano E<lt>kazeburo@gmail.comE<gt>

=cut



lib/WWW/Form/UrlEncoded/XS.pm
package WWW::Form::UrlEncoded::XS;

use strict;
use warnings;
use base qw/Exporter/;

our $VERSION = 0.04;
our @EXPORT_OK = qw/parse_urlencoded build_urlencoded/;

use XSLoader;
XSLoader::load(__PACKAGE__, $VERSION);


1;

__END__

=encoding utf-8

=head1 NAME

WWW::Form::UrlEncoded::XS - XS implementation of parser and builder for application/x-www-form-urlencoded

=head1 SYNOPSIS

    use WWW::Form::UrlEncoded::XS qw/parse_urlencoded build_urlencoded/;
    
    my $query_string = "foo=bar&baz=param";
    my @params = parse_urlencoded($query_string);
    # ('foo','bar','baz','param')
    
    my $query_string = build_urlencoded('foo','bar','baz','param');
    # "foo=bar&baz=param";


=head1 DESCRIPTION

WWW::Form::UrlEncoded::XS provides application/x-www-form-urlencoded parser and builder 
that is implemented by XS. see L<WWW::Form::UrlEncoded>'s document.

=head1 LICENSE

Copyright (C) Masahiro Nagano.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Masahiro Nagano E<lt>kazeburo@gmail.comE<gt>

=cut



lib/WWW/Form/UrlEncoded.pm
package WWW::Form::UrlEncoded;

use 5.008001;
use strict;
use warnings;

BEGIN {
    our $VERSION = "0.21";
    our @EXPORT_OK = qw/parse_urlencoded parse_urlencoded_arrayref build_urlencoded build_urlencoded_utf8/;

    my $use_pp = $ENV{WWW_FORM_URLENCODED_PP};

    if (!$use_pp) {
        eval {
            require WWW::Form::UrlEncoded::XS;
            if ( $WWW::Form::UrlEncoded::XS::VERSION < $VERSION ) {
                warn "WWW::Form::UrlEncoded::XS $VERSION is require. fallback to PP version";
                die;
            }
        };
        $use_pp = !!$@;
    }

    if ($use_pp) {
        require WWW::Form::UrlEncoded::PP;
        WWW::Form::UrlEncoded::PP->import(@EXPORT_OK);
    }
    else {
        WWW::Form::UrlEncoded::XS->import(@EXPORT_OK);
    }

    require Exporter;
    *import = \&Exporter::import;
}

1;
__END__

=encoding utf-8

=head1 NAME

WWW::Form::UrlEncoded - parser and builder for application/x-www-form-urlencoded

=head1 SYNOPSIS

    use WWW::Form::UrlEncoded qw/parse_urlencoded build_urlencoded/;
    
    my $query_string = "foo=bar&baz=param";
    my @params = parse_urlencoded($query_string);
    # ('foo','bar','baz','param')
    
    my $query_string = build_urlencoded('foo','bar','baz','param');
    # "foo=bar&baz=param";

=head1 DESCRIPTION

WWW::Form::UrlEncoded provides application/x-www-form-urlencoded parser and builder.
This module aims to have compatibility with other CPAN modules like 
HTTP::Body's urlencoded parser.

This module try to use L<WWW::Form::UrlEncoded::XS> by default and fail to it, 
use WWW::Form::UrlEncoded::PP instead

=head2 Parser rules

WWW::Form::UrlEncoded parsed string in this rule.

=over 4

=item 1. Split application/x-www-form-urlencoded payload by C<&> (U+0026) or C<;> (U+003B)

=item 2. Ready empty array to store C<name> and C<value>

=item 3. For each divided string, apply next steps.

=over 4

=item 1. If first character of string is C<' '> (U+0020 SPACE), remove it.

=item 2. If string has C<=>, let B<name> be substring from start to first C<=>, but excluding first C<=>, and remains to be B<value>. If there is no strings after first C<=>, B<value> to be empty string C<"">. If first C<=> is first character of the string, let B<key> be empty string C<"">. If string does not have any C<=>, all of the string to be B<key> and B<value> to be empty string C<"">.

=item 3. replace all C<+> (U+002B) with C<' '> (U+0020 SPACE).

=item 4. unescape B<name> and B<value>. push them to the array.

=back

=item 4. return the array.

=back

=head2 Test data

  'a=b&c=d'     => ["a","b","c","d"]
  'a=b;c=d'     => ["a","b","c","d"]
  'a=1&b=2;c=3' => ["a","1","b","2","c","3"]
  'a==b&c==d'   => ["a","=b","c","=d"]
  'a=b& c=d'    => ["a","b","c","d"]
  'a=b; c=d'    => ["a","b","c","d"]
  'a=b; c =d'   => ["a","b","c ","d"]
  'a=b;c= d '   => ["a","b","c"," d "]
  'a=b&+c=d'    => ["a","b"," c","d"]
  'a=b&+c+=d'   => ["a","b"," c ","d"]
  'a=b&c=+d+'   => ["a","b","c"," d "]
  'a=b&%20c=d'  => ["a","b"," c","d"]
  'a=b&%20c%20=d' => ["a","b"," c ","d"]
  'a=b&c=%20d%20' => ["a","b","c"," d "]
  'a&c=d'       => ["a","","c","d"]
  'a=b&=d'      => ["a","b","","d"]
  'a=b&='       => ["a","b","",""]
  '&'           => ["","","",""]
  '='           => ["",""]
  ''            => []

=head1 FUNCTION

=over 4

=item @param = parse_urlencoded($str:String)

parse C<$str> and return Array that contains key-value pairs.

=item $param:ArrayRef = parse_urlencoded_arrayref($str:String)

parse C<$str> and return ArrayRef that contains key-value pairs.

=item $string = build_urlencoded(@param)

=item $string = build_urlencoded(@param, $delim)

=item $string = build_urlencoded(\@param)

=item $string = build_urlencoded(\@param, $delim)

=item $string = build_urlencoded(\%param)

=item $string = build_urlencoded(\%param, $delim)

build urlencoded string from B<param>. build_urlencoded accepts arrayref and hashref values.

  build_urlencoded( foo => 1, foo => 2);
  build_urlencoded( foo => [1,2] );
  build_urlencoded( [ foo => 1, foo => 2 ] );
  build_urlencoded( [foo => [1,2]] );
  build_urlencoded( {foo => [1,2]} );

If C<$delim> parameter is passed, this function use it instead of using C<&>.

=item $string = build_urlencoded_utf8(...)

This function is almost same as C<build_urlencoded>. build_urlencoded_utf8 call C<utf8::encode> for all parameters.

=back

=head1 ENVIRONMENT VALUE

=over 4

=item * WWW_FORM_URLENCODED_PP

If true, WWW::Form::UrlEncoded force to load WWW::Form::UrlEncoded::PP.

=back

=head1 SEE ALSO

CPAN already has some application/x-www-form-urlencoded parser modules like these.

=over 4

=item L<URL::Encode>

=item L<URL::Encode::XS>

=item L<Text::QueryString>

=back

They does not fully compatible with WWW::Form::UrlEncoded. Handling of empty key-value
and supporting separator characters are different.

=head1 LICENSE

Copyright (C) Masahiro Nagano.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Masahiro Nagano E<lt>kazeburo@gmail.comE<gt>

=cut

